Misc updates

SKalt opened the pull request:
Hi @nickjj, I made several changes:
- improved unit test code coverage
- added PEP 484 type hints in stub files and complied with PEP 518 to include the in the package.
- added front- and back- end warning loggers of missing assets (closes #28 )
- allowed ingestion of flat asset maps  (may close #19 )
- eliminated the need to mark templates `| safe` in jinja templates by returning `jinja2.Markup` instead of raw strings
- allowed passing arbitrary key-value pairs to `javascript_tag`, `stylesheet_tag` (closes #15 )
- added webpack 4 examples including code splitting.
- misc documentation

nfvs committed:
`WEBPACK_MANIFEST_ASSETS_ONLY` - support simple manifest files.

Adds support for manifest files output with webpack-manifest-plugin
(https://www.npmjs.com/package/webpack-manifest-plugin)

nfvs committed:
Fix flake8 problems.

Commit:
Merge remote-tracking branch 'nvfs/flat_manifest' into warn_instead_of_failing

Commit:
Style fixes using `black --line-length=80`

SKalt committed:
documentation bump

SKalt committed:
Added warnings for missing assets, and a WEBPACK_LOG_LEVEL config

SKalt committed:
Log ioError rather than failing
Useful for testing APIs or anything that does not require JS bundles 
present.

SKalt committed:
kill the app if log_level == Error

SKalt committed:
moved type hinting to a stub file

SKalt committed:
Rename the test app to reflect it uses WP@^1

SKalt committed:
Stub cleanup: avoid explicit Any

SKalt committed:
Adding missing functions, fixing formatting errors

SKalt committed:
Adopted pyre-check for static type checking

SKalt committed:
flake8 fix

SKalt committed:
Added unit tests

SKalt committed:
made stubs distributable

SKalt committed:
Added pytest_cache to gitignore

SKalt committed:
Moved tests to more idiomatic ./tests directory

SKalt committed:
older tests are now working

SKalt committed:
Added type checking to the CI suite

SKalt committed:
ignored node_modules

SKalt committed:
More testing

SKalt committed:
completed asset map loading unit tests

SKalt committed:
Formatting [black]

SKalt committed:
Added *requirements.txt, needed for development, missing test fixtures

SKalt committed:
Added pep 518 pyproject.toml

SKalt committed:
More documentation, formatting

SKalt committed:
Updated README.md docs

SKalt committed:
Flake8 fix

SKalt committed:
Added documentation for development

SKalt committed:
added html sanitization function

germn commented:
Calling `url_for` with invalid param in Flask leads to `BuildError` exception.
Shouldn't `asset_url_for` and tags functions behave same way?

nickjj commented:
Wow, thanks a lot for all of those contributions.

I would have to do a deep dive code review before merging this, which may take a while.

Just popping in to say I've acknowledged your work.

@germn I like that idea.

SKalt commented:
@germn ditto-- that seems like the idiomatic way to do things! 

@nickjj I just looked through the Travis logs. 
I'll need to do some fixes on the BCC job 
(1)Thee3.5- jobs are failing due to `pyre` not supporting those versions. (2) the later jobs are failing due to my forgetting to install requirements.txt and devRequirements.txt. 
EDIT: done! The checks are more robust now.

SKalt committed:
Attempt to fix the travis jobs
Based of tornado's method of filtering by python versions

SKalt committed:
throw werkzeug.routing.BuildError on missing assets
outside of debugging contexts

SKalt committed:
Patched a final dict-order bug in the tests

SKalt commented:
Hi @nickjj, what's the status of your review?

nickjj commented:
Hi,

Sorry on the delay. I'm in heads down crazy work mode towards getting my next course released.

I will check it on this once that ships, but it may take a little while for that.

Definitely not ignoring your contributions!

SKalt committed:
Bump version to 0.2.0

Assuming everything's accepted in the flask-webpack PR, this will warrant at least a minor version bump.

SKalt committed:
fixed possibility of formatting "None" into urls

SKalt committed:
Merge branch 'master' of https://github.com/SKalt/flask-webpack

SKalt committed:
Remove inline styles: GitHub strips them

Moving some of the top-level sections from `<details>` to `##` is a TODO as well, but it looks fine for now.

SKalt committed:
fixed uncommited test fixtures

SKalt committed:
added support for chunked assets (unit tests passing)

SKalt committed:
adopted the `black` .flake8 file

SKalt committed:
fixed test, plus minor formatting from black

SKalt committed:
Merge pull request #1 from SKalt/accommodate-chunk-splitting

Accommodate chunk splitting

SKalt committed:
always check the given asset name first

SKalt committed:
Merge branch 'accommodate-chunk-splitting'

SKalt committed:
bump patch version due to change in the asset name resolution algorithm
see d2a577e

SKalt committed:
used jinja2 context to dedupe asset loading within a template

SKalt committed:
Merge branch 'accommodate-chunk-splitting'

SKalt committed:
version -> 0.2.2:
- assets are now deduped by default in the javscript_, stylesheet_tag
templates
- warning tags are now jinja2.Markup

SKalt commented:
For what it's worth, my team's now using this fork in production.  I'll see about downgrading the python 3 syntax (`{**dict_1, **dict_2}`) that travis's complaining about.

nickjj commented:
Wow thanks for so many patches. They haven't been going unnoticed. I need to set aside some time to get this merged in. If you can get it working with Python 2.x and 3.x that would be great.

SKalt commented on flask_webpack/__init__.py:
```diff
@@ -1,73 +1,212 @@
+        _attrs = {"_attrs": _attrs}
+    return " ".join(
+        key if value is True else '{}="{}"'.format(key, _escape(str(value)))
+        for key, value in ({**_attrs, **more_attrs}).items()
```
polyfill nicer syntax with
```python
def _merge(*dicts):
    return {key: value for key, value in d.items() for d in dicts}
```

SKalt committed:
dropped ** syntax

SKalt committed:
Dropped the last of the syntax incompatible with python 2.7

SKalt committed:
fixed type check

SKalt commented on flask_webpack/__init__.py:
```diff
@@ -1,73 +1,235 @@
+                "javascript_tag": self.javascript_tag,
+                "stylesheet_tag": self.stylesheet_tag,
+                "asset_urls_for": self.asset_urls_for,
             }
```
bug: `asset_url_for` registered only on flask >= 0.10

SKalt commented on flask_webpack/__init__.py:
```diff
@@ -1,73 +1,235 @@
+                self.assets = stats.get("assets") or stats
+            except IOError:
+                message = (
+                    "[Flask-Webpack] WEBPACK_MANIFEST_PATH='{}' must point to"
```
might want to name the bad path here.

SKalt commented:
> I need to set aside some time to get this merged in. 

Hey @nickjj, would you be willing to review this PR within a month? If not, I'd be willing to create a proper fork so that I get to experiment with pypi deployment tooling.
